### PR TITLE
(feature) Captcha service

### PR DESCRIPTION
**Pull request by Blaumaus**
Ticket: https://github.com/Swetrix/roadmap/issues/123

### CODE REVIEW

File: src/captcha/captcha.controller.ts [hunk 1]
```diff
@@ -9,4 +13,22 @@ export class CaptchaController {
   constructor(
     private readonly captchaService: CaptchaService,
   ) { }
+
+
+  @Post('/auto-verify')
+  @UseGuards(BotDetectionGuard)
+  @BotDetection()
```
Comment (yevheniikulisidi) [inline src/captcha/captcha.controller.ts:50]: You can make a combined decorator - https://docs.nestjs.com/custom-decorators#decorator-composition

File: src/captcha/captcha.service.ts [hunk 2]
```diff
@@ -1,10 +1,43 @@
...
 } from '@nestjs/common'
+import * as svgCaptcha from 'svg-captcha'
+import { hash } from 'blake3'
+
+import { CAPTCHA_SALT } from '../common/constants'
+
+interface GeneratedCaptcha {
+  data: string
+  hash: string
+}
```
Comment (yevheniikulisidi) [inline src/captcha/captcha.service.ts:29]: The best solution would be to put the interface in a separate file.

**Review by yevheniikulisidi — CHANGES_REQUESTED** [review]

File: src/captcha/dtos/manual.dto.ts [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+  hash?: string
+
+  @ApiProperty({
+    required: true,
+    example: '4vic2',
+    description: 'Captcha code',
+  })
+  @IsNotEmpty()
+  code?: string
+}
```
Comment (yevheniikulisidi) [inline src/captcha/dtos/manual.dto.ts]: Add stricter validation. You can use this table - https://github.com/typestack/class-validator#validation-decorators

**Review by yevheniikulisidi — CHANGES_REQUESTED** [review]

File: src/captcha/dtos/generate.dto.ts [hunk 4]
```diff
@@ -8,5 +9,6 @@ export class GenerateDTO {
     required: false,
     description: 'Captcha theme',
   })
+  @IsOptional()
   theme?: string
```
Comment (yevheniikulisidi) [inline src/captcha/dtos/generate.dto.ts]: More validation!

File: src/captcha/dtos/validate.dto.ts [hunk 5]
```diff
@@ -20,6 +20,7 @@ export class ValidateDTO {
     required: false,
     description: 'Captcha hash (if available)',
   })
+  @IsOptional()
   hash?: string
```
Comment (yevheniikulisidi) [inline src/captcha/dtos/validate.dto.ts]: More validation!

File: src/project/dto/project.dto.ts [hunk 6]
```diff
@@ -46,4 +46,10 @@ export class ProjectDTO {
       "When true, anyone on the internet (including Google) would be able to see the project's Dashboard.",
   })
   public: boolean
+
+  @ApiProperty({
+    required: false,
+    description: 'When true, created a new Captcha Project without Analytics Projects.'
+  })
+  isCaptcha: boolean
```
Comment (yevheniikulisidi) [inline src/project/dto/project.dto.ts:54]: There is no validation at all.

File: src/project/entity/project.entity.ts [hunk 7]
```diff
@@ -69,4 +69,13 @@ export class Project {
...
   extensions: ExtensionToProject[]
+
+  @Column({ default: true })
+  isAnalyticsProject: boolean
+
+  @Column({ default: false})
+  isCaptchaProject: boolean
+
+  @Column({ default: false})
+  isCaptchaEnabled: boolean
```
Comment (yevheniikulisidi) [inline src/project/entity/project.entity.ts]: Add an entity column type.

File: src/project/project.controller.ts [hunk 8]
```diff
@@ -103,10 +106,17 @@ export class ProjectController {
       const where = Object()
       where.admin = userId
 
+      if (isCaptcha) {
+        where.isCaptchaProject = true
+      } else {
+        where.isAnalyticsProject = true
+      }
```
Comment (yevheniikulisidi) [inline src/project/project.controller.ts:114]: It's just terrible code. No typing, and you can use the conditional operator here.

**Review by yevheniikulisidi — CHANGES_REQUESTED** [review]

**Review by yevheniikulisidi — CHANGES_REQUESTED** [review]

**Review by yevheniikulisidi — CHANGES_REQUESTED** [review]

File: src/captcha/dtos/validate.dto.ts [hunk 9]
```diff
@@ -15,18 +15,4 @@ export class ValidateDTO {
   })
   @IsNotEmpty()
   secret: string
```
Comment (yevheniikulisidi) [inline src/captcha/dtos/validate.dto.ts]: Add validation.

**Review by yevheniikulisidi — CHANGES_REQUESTED** [review]

File: src/project/entity/project.entity.ts [hunk 10]
```diff
@@ -48,6 +48,24 @@ export class Project {
...
+  @Column({ default: true })
+  isAnalyticsProject: boolean
+
+  @ApiProperty()
+  @Column({ default: false })
+  isCaptchaProject: boolean
+
+  @ApiProperty()
+  @Column({ default: false })
+  isCaptchaEnabled: boolean
```
Comment (yevheniikulisidi) [inline src/project/entity/project.entity.ts:63]: Add an entity column type.

**Review by yevheniikulisidi — CHANGES_REQUESTED** [review]

File: src/captcha/captcha.controller.ts (hunk 1 above)
Comment (Blaumaus) [reply src/captcha/captcha.controller.ts:50]: in the future

File: src/captcha/captcha.controller.ts [hunk 11]
```diff
@@ -14,21 +20,46 @@ export class CaptchaController {
     private readonly captchaService: CaptchaService,
   ) { }
 
-
-  @Post('/auto-verify')
+  @Get('/auto-verifiable')
```
[the remark this answers is not in the document]
Comment (Blaumaus) [reply src/captcha/captcha.controller.ts]: it's used in other places like project, user, etc. so I'll keep it to preserve code consistency

**Review by yevheniikulisidi — CHANGES_REQUESTED** [review]
The functions must be inside the `CaptchaService` class.

**Review by yevheniikulisidi — CHANGES_REQUESTED** [review]

**Pull request merged**